pkg/api/grpc/grpc.go

```diff
@@ -426,16 +441,19 @@ func (a *api) BulkPublishEventAlpha1(ctx context.Context, in *runtimev1pb.BulkPu
 	bulkRes := runtimev1pb.BulkPublishResponse{}
 
 	if err != nil {
-		// Only respond with error if it is  permission denied or not found.
+		var nerr error
+		// Only respond with error if it is permission denied or not found.
 		// On error, the response will be empty.
-		nerr := status.Errorf(codes.Internal, messages.ErrPubsubPublishMessage, topic, pubsubName, err.Error())
+		nerr = apierrors.PubSubPublishMessage(pubsubName, string(contribMetadata.PubSubType), topic, err)
+
 		if errors.As(err, &runtimePubsub.NotAllowedError{}) {
-			nerr = status.Errorf(codes.PermissionDenied, err.Error())
+			nerr = apierrors.PubSubPublishForbidden(pubsubName, string(contribMetadata.PubSubType), topic, a.AppID(), err)
 		}
 
 		if errors.As(err, &runtimePubsub.NotFoundError{}) {
-			nerr = status.Errorf(codes.NotFound, err.Error())
+			nerr = apierrors.PubSubTestNotFound(pubsubName, string(contribMetadata.PubSubType), topic, err)
 		}
+
```

Comment:
Nit: please use a switch statement for readability with the 3 cases:

1. `errors.As(err, &runtimePubsub.NotAllowedError{})`
2. `errors.As(err, &runtimePubsub.NotFoundError{})`
3. default

Also, shouldn't this be using `errors.Is` rather than `As`, since you're not reading from the "casted" error?

Reply:
I didn't author that code. I can update it as suggested to use `errors.Is()`

Reply:
done

Reply:
Actually, switching from errors.As -> errors.Is was causing test failures. I'm switching back and keeping it as `errors.As`